test/gtest/uct/test_flush.cc

```diff
@@ -538,3 +539,255 @@ UCS_TEST_SKIP_COND_P(uct_flush_test, am_pending_flush_nb,
+        if (packed_len >= 0) {
+            EXPECT_EQ(sendbuf.length(), (size_t)packed_len);
+            return UCS_OK;
+        } else {
+            return (ucs_status_t)packed_len;
+        }
+    }
+
+    ucs_status_t am_zcopy(side *s) {
+        mapped_buffer &sendbuf = *s->buf1k;
+        ucs_status_t status;
+        size_t header_length = 0;
+        uct_iov_t iov;
+
+        iov.buffer = (char*)sendbuf.ptr() + header_length;
+        iov.count  = 1;
+        iov.length = sendbuf.length() - header_length;
+        iov.memh   = sendbuf.memh();
+        status = uct_ep_am_zcopy(s->e->ep(0), 0, sendbuf.ptr(), header_length,
+                                 &iov, 1, 0, NULL);
+        return status;
+    }
+
+    ucs_status_t get_bcopy(side *s) {
+        const mapped_buffer& sendbuf = *s->buf32;
+        const mapped_buffer& recvbuf = *s->peer->buf32;
+
+        return uct_ep_get_bcopy(s->e->ep(0), (uct_unpack_callback_t)memcpy,
+                                sendbuf.ptr(), sendbuf.length(),
+                                recvbuf.addr(), recvbuf.rkey(), NULL);
+    }
+
+    struct comp : public uct_completion_t {
+        comp() : flag(false) {
+            count = 0;
+            func  = comp_cb;
+        }
+
+        volatile bool flag;
+        static void comp_cb(uct_completion_t *self, ucs_status_t status) {
+            comp *c = static_cast<comp*>(self);
+            ASSERT_UCS_OK(status);
+            c->flag = true;
+        }
+    };
+
+    void flush_and_reconnect() {
+        std::list<entity *> flushing;
+        ucs_status_t status = UCS_OK;
+        struct comp done;
+
+        flushing.push_back(s0.e);
+        flushing.push_back(s1.e);
+        done.count = flushing.size();
+        ucs_time_t loop_end_limit = ucs_get_time() + ucs_time_from_sec(50.0);
+        while (ucs_get_time() < loop_end_limit) {
+            std::list<entity *>::iterator i = flushing.begin();
+            while (i != flushing.end()) {
+                status = uct_ep_flush((*i)->ep(0), UCT_FLUSH_FLAG_CANCEL, &done);
+                if (status == UCS_ERR_NO_RESOURCE) {
+                    i++;
+                } else {
+                    ASSERT_UCS_OK_OR_INPROGRESS(status);
+                    i = flushing.erase(i);
+                    if (status == UCS_OK) {
+                        done.count--;
+                    }
+                }
+            }
+
+            if (flushing.empty()) {
+                break;
+            }
+
+            while (progress()) {}
+            usleep(100);
+        }
+        ASSERT_UCS_OK_OR_INPROGRESS(status);
+
+        if (done.count) {
+            wait_for_flag(&done.flag);
+        }
+
+        s1.e->destroy_eps();
+        s1.e->connect(0, *s0.e, 0);
+    }
+
+    typedef ucs_status_t (uct_cancel_test::* send_func_t)(side *s);
+
+    void fill(send_func_t send) {
+        ucs_status_t status;
+        std::list<side *> filling;
+
+        filling.push_back(&s0);
+        filling.push_back(&s1);
```

Comment:
maybe add fill() to "side" class?

Reply:
here we fill sides simultaneously

Reply:
is it important for the test case to fill simultaneously? as long as there is no progress() in between?
IMO it's important only for flush

Reply:
we want to call flush having both TX and RX operations in-flight